### PR TITLE
Add P&E Mailer module pe-mailer

### LINKED ISSUE #7: Connect cyhy customer db to encrypt and embed pdf reports.
# 💡 Summary #
Update the "report_generator.py" scripts to include a function "generate_reports" that coordinates the P&E Report process.

## Motivation and context ##
This is a key feature to securely deliver encrypted reports to Posture and Exposure customers.

## Implementation notes ##

Please provide details for implementation, such as:

- Create function "generate_reports" to include the following:
- Connect to the cyhy database. 
- Gathers customer information based on customer ID. 
- Builds the reports.
- Embed raw data into the pdf output. 
- Encrypts the pdfs.

## Acceptance criteria ##

How do we know when this work is done?

- [x] Produce P&E Report with no errors.
- [x] Output is encrypted.
- [x] Validate report data.
- [x] Review and ensure reports can be sent using current mailer scripts.

### LINKED ISSUE #8: Add command line functionality.
# 💡 Summary #

Add in command line functionality that addresses cyhy database connection, data input and data output arguments.

## Motivation and context ##

Critical feature to ensure a secure and proper distribution of P&E Reports.

## Implementation notes ##

Intended usage should follow:

```
"""A tool for creating Posture & Exposure reports.

Usage:
    pe-reports REPORT_DATE DATA_DIRECTORY OUTPUT_DIRECTORY [--db-creds-file=FILENAME] [--log-level=LEVEL]

Arguments:
  REPORT_DATE                   Date of the report, format YYYY-MM-DD.
  DATA_DIRECTORY                The directory where the excel data files are located.
                                Organized by owner.
  OUTPUT_DIRECTORY              The directory where the final PDF reports should be saved.
  -c --db-creds-file=FILENAME   A YAML file containing the Cyber
                                Hygiene database credentials.
                                [default: /secrets/database_creds.yml]

Options:
  -h --help                     Show this message.
  -v --version                  Show version information.
  --log-level=LEVEL             If specified, then the log level will be set to
                                the specified value.  Valid values are "debug", "info",
                                "warning", "error", and "critical". [default: info]
"""
```

## Acceptance criteria ##

How do we know when this work is done?

- [x] Successfully generate P&E Reports from command line interface.
- [x] DATA_DIRECTORY input data file locations are not hard-coded.
- [x] OUTPUT_DIRECTORY is created when scripts are run.
- [x] Proper Exceptions are made if -db-creds-file is not available or fails.
- [x] log-level option type is added.

### LINKED ISSUE #17: Add P&E Mailer Automation - Module
# 💡 Summary #

To further automate the delivery of P&E Reports to customers, a module/program will be implemented to retrieve pe_report output and deliver to the appropriate customer's email. Ideally this would be an in-line process that can be turned on/off will running the pe_reports module.

## Motivation and context ##

This would fully automate the P&E Reports delivery process. Secondarily allows for scalability and acquiring more customers at a faster pace. With proper test, it improves quality control.

## Implementation notes ##

- Proposed Architecture

```
├── src
│   ├── pe_reports
│   │   ├── __init__.py
│   │   ├── __main__.py
│   │   ├── _version.py
│   │   ├── data
│   │   ├── file_1.py
│   │   ├── file_2.py
│   │   └── report_generator.py
│   ├── pe_mailer
│   │   ├── __init__.py
│   │   ├── __main__.py
│   │   ├── _version.py
│   │   ├── data
│   │   ├── file_1.py
│   │   ├── file_2.py
│   │   └── mailer.py
└── tests
    ├── conftest.py
    └── test_pe_reports.py
    └── test_pe_mailer.py
├── setup.py
├── docker-compose/yml????
```
- Elegantly include secondary command line process for [mailer Usage](https://github.com/cisagov/cyhy-mailer).
- Update setup.py  to handle multiple `entry_points={'console_scripts':[ ]`
- Update dependancies for a python `import pe_mailer`

## Acceptance criteria ##

How do we know when this work is done?

- [x] TBD

### LINKED ISSUE #50: Define StatsMessage in pe-mailer module
# 🐛 Summary #

There is a definition missing in `email_reports.py` called `StatsMessage`. While PR is in draft mode, the `StatsMessage` has been temporarily hardcode as a string. 

line 442
```
if summary_to is not None and all_stats_strings:
        StatsMessage = "Needs Defined!!"
        message = StatsMessage(summary_to.split(","), all_stats_strings)
        try:

```

## Expected behavior ##

`StatsMessage` will have a reference to src/pe_mailer/StatsMessage.py which is originally taken from [cyhy-mailer](https://github.com/cisagov/cyhy-mailer/blob/develop/cyhy/mailer/StatsMessage.py).

### LINKED ISSUE #51: Reestablish args for pande_dir and db_creds_file in pe-mailer module
# 💡 Summary #
Currently the arguments `pande_dir `and `db_creds_file` in the pe-mailer module are hard-coded only for the reson to pass checks for the draft PR. 

## Motivation and context ##

This would be useful because... It is the primary command-line to invoke the pe-mailer module.

## Implementation notes ##

- Reference `pande_dir `and `db_creds_file`arg refereces to the arguments in main()

## Acceptance criteria ##

- [ ] Usage should be: `% pe-mailer pande_report_dir= db_creds_file= summary_to=None test_emails= debug=None`

### LINKED ISSUE #52: mypy hook: PyYAML not supported with Python 3.9
# 🐛 Summary #

During run of `pre-commit run --all-flies` mypy check issues the following warning: "PyYAML not supported with Python 3.9"

## To reproduce ##

1. Have `import yaml ` in scripts. [PyPi Reference](https://pypi.org/project/PyYAML/)
2. `pre-commit run --all-flies`

## Expected behavior ##

- pre-commit would pass after implementing the following parameter to `mypy` hook - additional_dependencies: [types-all]
- Reference: https://github.com/asottile/types-all

### LINKED ISSUE #53: Update pe-mailer error handling
# 💡 Summary #

Include the proper "Exceptions" and "Logging" to handle errors that may cause the pe-mailer module "Fails and Traceback messaging". On run `pe-mailer ... ... ...` the module should run start-to-finsh regards less of a lack of module resources, or return a "Usage" error message. 

For a code example see [pe-reports fix and its scope](https://github.com/cisagov/pe-reports/pull/6#issuecomment-858723415).

## Motivation and context ##

It provides a proper implementation of the pe-mailer module, and handles errors properly when importing and running.

## Implementation notes ##

Please provide details for implementation, such as:

- Explore all probable Traceback Issues
- Implement Exceptions to issues with proper messaging. 

## Acceptance criteria ##
- [x] `% import pe-mailer`
- [x] `% pe-mailer pande_report_dir= db_creds_file= ` with bad references but runs to completion with no Traceback errors.

### LINKED ISSUE #54: Update README.md (develop)
# 💡 Summary #

Update README.md content in association to repo changes which may include the following items:
- Removal of LibreOffice dependancies
- Update to pe-reports cmd-line usage
- Updates addressing the pe-mailer module
- Updates to install process
- Description on report generation and delivery processes.
- Description of PE Report metrics.   

## Motivation and context ##

Necessary to provide appropriate instruction for using repository resources and defining its purpose.

## Acceptance criteria ##

- [ ] P&E Team Sign-Off (all members)
- [ ] Commit README.md file in pull request

### LINKED ISSUE #62: Add cc Address To CyHy Mailer Module
# 🐛 Summary #

After talking to Genevieve, we need to ensure that the CyHy mailer sends a copy of all reports to reports@cyber.dhs.gov

## Expected behavior ##

All reports from this point forward should cc reports@cyber.dhs.gov for record retention requirements.

### PR DESCRIPTION
**Pull request by schmelz21**
## 🗣 Description ##

The PE Mailer Module (pe-mailer) is intended to be a secondary process that delivers P&E Reports generated by the "pe-reports" module.  The user, (P&E Team Member) runs a command that gathers the P&E Reports "generated" for each customer and sends out an encrypted document via email to each. 

## 💭 Motivation and context ##
The motivation is to allow any P&E Team Member the ability to  "securely" deliver the bi-weekly P&E Reports. Secondary motives include that the process is automated, it improves quality assurance and sets the cisagov/pe-reports to be easily scalable for future customers.

PR related Issue= https://github.com/cisagov/pe-reports/issues/17

Dependancies include:
- Access to customer database - [mongo-db-from-config](https://github.com/cisagov/mongo-db-from-config/)
- CISA AWS SES capabilities

## 🧪 Testing ##
The following processes must run successfully
Install: `pip install -e .  `
Tests: `pytest /tests`

Additionally pass all pre-commits
`pre-commit run --all-files  `

Run pe-reports
`pe-reports REPORT_DATE DATA_DIRECTORY OUTPUT_DIRECTORY [--db-creds-file=FILENAME] [--log-level=LEVEL]`

Run pe-mailer
`pe-mailer pande_report_dir= db_creds_file= summary_to=None, test_emails= debug=None`

## ✅ Checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All_ Active TODOs are addressed in this PR.
- [x]  Update [README.md](https://github.com/cisagov/pe-reports/blob/develop/README.md)
- [x] Address any LGTM alerts.
- [x] _All_ future TODOs are captured in issues, which are referenced
      in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.

### CODE REVIEW

File: README.md
Comment (dav3r) [inline README.md]: The phrase "raw data" is not normally hyphenated:
```suggestion
collects the raw data and creates the encrypted PDFs. The `pe_mailer` then
```

File: README.md
Comment (dav3r) [inline README.md:7]: This `snyk` badge is part of our [skeleton](https://github.com/cisagov/skeleton-python-library/blob/develop/README.md#skeleton-python-library) and should not be removed.  Though FYI, this badge may be removed from the skeleton in the future.

**Review by cduhn17 — APPROVED** [review]
At READ.md the Link at Load[ AWS Profile](https://github.com/cisagov/cool-dns-cyber.dhs.gov) returned 404.

**Comment by schmelz21** [comment]
> At READ.md the Link at Load[ AWS Profile](https://github.com/cisagov/cool-dns-cyber.dhs.gov) returned 404.

@cduhn17 I will take a look at the broken README.md link. 

In the meantime can you take a look at this question I put out to the group regarding Usage. I want to build some continuity between cmd lines. I have this update separate from this PR, but if you think it is worth the effort, I could update in this PR and link the issue. Ping me with questions...

Proposed - https://github.com/cisagov/pe-reports/issues/54#issuecomment-926128289

File: README.md
Comment (dav3r) [inline README.md]: This is more accurate:
```suggestion
- Configure [cisagov MongoDB connection](https://github.com/cisagov/mongo-db-from-config)
```

File: README.md
Comment (dav3r) [inline README.md]: Typo:
```suggestion
```console
```

File: README.md
Comment (dav3r) [inline README.md]:
```suggestion
  DATA_DIRECTORY                The directory where the Excel data files are located.
```

File: README.md
Comment (dav3r) [inline README.md]: Typo:
```suggestion
```console
```

File: README.md
Comment (dav3r) [inline README.md]: Same as above:
```suggestion
- Configure [cisagov MongoDB connection](https://github.com/cisagov/mongo-db-from-config)
```

File: README.md
Comment (dav3r) [inline README.md]: I'm not sure what you mean by "Load AWS Profile" here.  The `cisagov/cool-dns-cyber.dhs.gov` repository contains Terraform code for configuring a DNS zone and related IAM roles.  How does that apply here?

File: setup.py
Comment (dav3r) [inline setup.py]: Please create this issue and add a comment with the issue URL.

File: setup.py
Comment (dav3r) [inline setup.py]: This line breaks the alphabetical order of this list - please correct it.

File: .pre-commit-config.yaml
Comment (dav3r) [inline .pre-commit-config.yaml]: Prefer alphabetical order:
```suggestion
          - types-PyYAML
          - types-requests
```

File: src/pe_mailer/__init__.py
Comment (dav3r) [inline src/pe_mailer/__init__.py]: Why is this comment here?  It seems to have been copied from somewhere else, but it doesn't seem to apply.

File: src/pe_mailer/_version.py
Comment (dav3r) [inline src/pe_mailer/_version.py]: There's no need to change this from the skeleton:
```suggestion
"""This file defines the version of this module."""
```

File: src/pe_mailer/message.py
Comment (dav3r) [inline src/pe_mailer/message.py:38]: Just double-checking that all of these email addresses are correct for P&E reports?

File: src/pe_mailer/pe_message.py
Comment (dav3r) [inline src/pe_mailer/pe_message.py]: This file should be updated to use [`chevron`](https://github.com/noahmorrison/chevron) instead of `pystache`.  See https://github.com/cisagov/cyhy-mailer/pull/90 for details on how and why we recently did that in `cyhy-mailer`.

File: src/pe_mailer/pe_message.py
Comment (dav3r) [inline src/pe_mailer/pe_message.py]:
```suggestion
            report attachment. Use dates like "December 12, 2017".
```

File: src/pe_mailer/pe_message.py
Comment (dav3r) [inline src/pe_mailer/pe_message.py]: These should be changed to `chevron` - see previous comment about this.

File: src/pe_mailer/stats_message.py
Comment (dav3r) [inline src/pe_mailer/stats_message.py]: `pystache` -> `chevron` again.

File: src/pe_mailer/stats_message.py
Comment (dav3r) [inline src/pe_mailer/stats_message.py]:
```suggestion
    Subject = "pe-mailer summary from {{date}}"
```

File: src/pe_mailer/stats_message.py
Comment (dav3r) [inline src/pe_mailer/stats_message.py]:
```suggestion
Here is the pe-mailer summary from the run ending at {{date}}:
```

File: src/pe_mailer/stats_message.py
Comment (dav3r) [inline src/pe_mailer/stats_message.py]: Not sure what email address you want to use here:
```suggestion
Please direct feedback and questions to ???@??? and/or the pe-mailer GitHub project.

Regards,
The P&E Development Team
Cybersecurity and Infrastructure Security Agency (CISA)
???@???
```

File: src/pe_mailer/stats_message.py
Comment (dav3r) [inline src/pe_mailer/stats_message.py]:
```suggestion
Here is the pe-mailer summary from {{date}}:
```

File: src/pe_mailer/stats_message.py
Comment (dav3r) [inline src/pe_mailer/stats_message.py]: Same as above:
```suggestion
<p> Please direct feedback and questions to <a
href="mailto:???@???">the PE Development Team</a>
and/or the <a
href="https://github.com/cisagov/pe-reports">pe-reports GitHub
project</a>.</p>

<p>
Regards,<br>
The P&E Development Team<br><br>
Cybersecurity and Infrastructure Security Agency<br>
<a href="mailto:???@???">???@???</a>
```

File: src/pe_mailer/stats_message.py
Comment (dav3r) [inline src/pe_mailer/stats_message.py]: `pystache` -> `chevron`

File: src/pe_reports/_version.py
Comment (dav3r) [inline src/pe_reports/_version.py]: No need to change this from the skeleton:
```suggestion
"""This file defines the version of this module."""
```

File: src/pe_reports/report_generator.py
Comment (dav3r) [inline src/pe_reports/report_generator.py]:
```suggestion
  DATA_DIRECTORY                The directory where the Excel data files are located.
```

File: src/pe_reports/report_generator.py
Comment (dav3r) [inline src/pe_reports/report_generator.py]: Will this TODO still be valid after this PR is merged?  If not, it can be deleted.  Otherwise, it should stay.

File: src/pe_reports/report_generator.py
Comment (dav3r) [inline src/pe_reports/report_generator.py]: Not sure why this change was made, but it should be reverted:
```suggestion
def main() -> None:
```

**Review by dav3r — CHANGES_REQUESTED** [review]
Apologies for my delay in reviewing this.  Please take a look at the changes I requested and let me know if you have any questions.

File: setup.py
Comment (jsf9k) [inline setup.py]: You should be using chevron instead of pystache.  The latter is no longer supported and does not work with Python 3.9.

File: src/pe_mailer/pe_message.py
Comment (jsf9k) [inline src/pe_mailer/pe_message.py]: I'd suggest something slightly different than what @dav3r did:
```suggestion
            report attachment. Specify dates using the format "December 12, 2017".
```

**Review by jsf9k — CHANGES_REQUESTED** [review]

File: README.md
Comment (schmelz21) [reply README.md]: @dav3r before running pe-mailer cmd we have to load the SES profile via terminal. I wasn't sure how best to represent that step, or if to include it at all. Any recommendation around this particular step would be helpful.

File: setup.py
Comment (schmelz21) [reply setup.py]: @dav3r, @jsf9k I just want to confirm that the issue here is not duplicate to https://github.com/cisagov/pe-reports/issues/52, and that there is an alternative solution to be explored.

File: README.md
Comment (jsf9k) [reply README.md]: The link you added doesn't really clue anyone in as to what you mean by the "AWS profile".  I suggest something like this:
```markdown
- Load an AWS profile that assumes [this role](https://github.com/cisagov/cool-dns-cyber.dhs.gov/blob/develop/sessendemail_role.tf#L33-L39)
```

File: README.md
Comment (jsf9k) [reply README.md]: @schmelz21, please allow the reporter (in this case @dav3r) to review your changes and resolve the conversation.

Comment (jsf9k) [reply README.md]: Ah, nevermind.  I see that you committed @dav3r's suggestion. :man_facepalming:

Comment (schmelz21) [reply README.md]: yes ran a "suggestion batch" on the minor txt, code fixes.

File: README.md
Comment (schmelz21) [reply README.md:7]: @dav3r to confirm, I was getting a broken link response from SNYK... "The repository doesn't exist, or does not contain a relevant manifest file." If I put it back in is there a resolution I should be pursuing to that issue.

File: src/pe_mailer/message.py
Comment (schmelz21) [reply src/pe_mailer/message.py:38]: @stewartl97 can you confirm this for the team that this is proper configuration:

    DefaultFrom = "reports@cyber.dhs.gov"
    DefaultCc = None
    DefaultBcc = ["cyhy_reports@hq.dhs.gov", "reports@cyber.dhs.gov"]
    DefaultReplyTo = "vulnerability@cisa.dhs.gov"

File: README.md
Comment (dav3r) [reply README.md:7]: There is no immediate resolution for this issue - it's ok for this to remain broken for the moment.  I'm not familiar enough with snyk to understand why this is failing - perhaps @jsf9k or @mcdonnnj could take a look at this.  When I added `pe-reports` to our list of projects at snyk.io this was the output I saw:
```
Job started: 30 September 2021, 16:32:46
Processed cisagov/pe-reports from GitHub
Found Snyk Code supported files [develop] View project
Found requirements-dev.txt [develop] Failed to process manifest requirements-dev.txt
Found requirements-test.txt [develop] Failed to process manifest requirements-test.txt
Found requirements.txt [develop] Failed to process manifest requirements.txt
1 project created
Job completed: 30 September 2021, 16:33:04
```

Even though it says "1 project created", I still don't see `pe-reports` in our list of projects [here](https://app.snyk.io/org/cisagov/projects).

File: README.md
Comment (dav3r) [reply README.md]: @schmelz21 - I like @jsf9k's suggestion.

File: setup.py
Comment (dav3r) [reply setup.py]: I'm not sure and don't have time at the moment to confirm this.  If the comment is still correct, then there should be a corresponding issue linked.  If the comment is no longer correct, then it should be deleted.

File: src/pe_mailer/pe_message.py
Comment (dav3r) [reply src/pe_mailer/pe_message.py]: Also good - I'm not tied to my suggestion.

File: src/pe_reports/report_generator.py
Comment (schmelz21) [reply src/pe_reports/report_generator.py]: @dav3r - yes technically this will be resolved now, both here and in 'main' branch PRs. It has been removed.
c5ba70c

File: README.md
Comment (schmelz21) [reply README.md:7]: Added snyk.io link for now - commit c5ba70c

File: setup.py
Comment (schmelz21) [reply setup.py]: Updated - baacfb6

File: src/pe_mailer/stats_message.py
Comment (schmelz21) [reply src/pe_mailer/stats_message.py]: Updated - baacfb6

File: README.md
Comment (schmelz21) [reply README.md]: Updated in commit 1f042b6

File: src/pe_mailer/pe_message.py
Comment (schmelz21) [reply src/pe_mailer/pe_message.py]: Committed - baacfb6

File: src/pe_mailer/pe_message.py
Comment (schmelz21) [reply src/pe_mailer/pe_message.py]: Updated with this suggestions - ed0a1c7

File: src/pe_mailer/pe_message.py
Comment (schmelz21) [reply src/pe_mailer/pe_message.py]: Committed - baacfb6

File: src/pe_mailer/stats_message.py
Comment (schmelz21) [reply src/pe_mailer/stats_message.py]: Committed - baacfb6

File: src/pe_mailer/message.py
Comment (stewartl97) [reply src/pe_mailer/message.py:38]: This is correct.  It was requested that we Bcc the reports@cyber.dhs.gov address by Genevieve for records management, and cyhy_reports@hq.dhs.gov was already written into the code so we did not want to change that portion without approval.

File: src/pe_mailer/__init__.py
Comment (schmelz21) [reply src/pe_mailer/__init__.py]: Bad copy over - updated in commit: 6f8b5d8

File: src/pe_mailer/stats_message.py
Comment (schmelz21) [reply src/pe_mailer/stats_message.py]: Bad copy over - updated in commit: 6f8b5d8

File: src/pe_mailer/stats_message.py
Comment (schmelz21) [reply src/pe_mailer/stats_message.py]: Bad copy over - updated in commit: 6f8b5d8

File: setup.py
Comment (schmelz21) [reply setup.py]: Alphabetized in 4ca706f

File: setup.py
Comment (schmelz21) [reply setup.py]: @dav3r , I removed the comment and added back in python 3.9 due to the resolution in issue  #52. This is a miss on my part and should have been removed prior to the PR review request.
Updated in commit 4ca706f

File: README.md
Comment (jsf9k) [reply README.md:7]: It's probably because it's the end of the month and our Snyk credits have run out.

**Review by jsf9k — APPROVED** [review]

**Comment by schmelz21** [comment]
Hi @dav3r, are you able to schedule some time to re-review?
Thanks

File: src/pe_reports/report_generator.py
Comment (dav3r) [reply src/pe_reports/report_generator.py]: Great, thanks!

**Comment by dav3r** [comment]
> Hi @dav3r, are you able to schedule some time to re-review? Thanks

Yes, doing this now.  Sorry for the delay getting back to this- I was on leave for a few days.

File: setup.py
Comment (dav3r) [reply setup.py]: Looks good now. 👍

**Review by dav3r — APPROVED** [review]
This all looks solid now - nice work! 🦅

**Pull request merged**
